### PR TITLE
refactor(client): Convert durations into duration-js issue1657

### LINKED ISSUE #1657: Convert durations into a more readable format
via https://github.com/mozilla/fxa-content-server/pull/1655#discussion_r17499458

In a lot of places we have code like this:
```js
var CERT_DURATION_MS = 1000 * 60 * 60 * 6; // 6hrs
var ASSERTION_DURATION_MS = 1000 * 3600 * 24 * 365 * 25; // 25 years
```

It may be nice if we could switch to some duration library to do something a bit more readable like [**duration-js**](https://www.npmjs.org/package/duration-js):
```js
var CERT_DURATION_MS = new Duration('6h');
var ASSERTION_DURATION_MS = new Duration('52w') * 25; // 25 years
```

Or using [**moment**](https://www.npmjs.org/package/moment):
```js
var CERT_DURATION_MS = moment.duration(6, 'hours');
var ASSERTION_DURATION_MS = moment.duration(25, 'years');
```

---

Actually, here's a longer list of places where we're making up some times:
```sh
$ grep -Ir "1000 *" . --exclude-dir="node_modules" --exclude-dir=".tmp" --exclude-dir="app/bower_components" --exclude-dir="dist"  | grep -v ".svg:" | grep -v "bower_components" | grep -v app/scripts/vendor/

./app/scripts/lib/assertion.js:  var CERT_DURATION_MS = 1000 * 60 * 60 * 6; // 6hrs
./app/scripts/lib/assertion.js:  var ASSERTION_DURATION_MS = 1000 * 60 * 5; // 5mins
./app/scripts/lib/channels/fx-desktop.js:  var DEFAULT_SEND_TIMEOUT_LENGTH_MS = 1000;
./app/scripts/lib/metrics.js:  var TEN_MINS_MS = 10 * 60 * 1000;
./app/scripts/views/form.js:    LONGER_THAN_EXPECTED: 10000, // 10 seconds
./app/scripts/views/mixins/resend-mixin.js:  var SHOW_RESEND_IN_MS = 5 * 60 * 1000; // 5 minutes.
./app/tests/spec/views/confirm.js:            }, view.VERIFICATION_POLL_IN_MS + 1000);
./server/bin/csp-violation-server.js:      when: (new Date()).getTime() / 1000,
./server/lib/configuration.js:    default: 10 * 60 * 1000 // 10 minutes
./tests/functional/firefox/same_account.js:        .sleep(1000)
./tests/functional/lib/helpers.js:    }, [ selector ], 10000);
./tests/lib/restmail.js:          }, 1000);
./tests/server/configuration.js:    var dfd = this.async(10000);
./tests/server/metrics-collector-stderr.js:    var dfd = this.async(1000);
./tests/server/proxy.js:    var dfd = this.async(1000);
./tests/server/proxy.js:    var dfd = this.async(1000);
./tests/server/proxy.js:    var dfd = this.async(1000);
./tests/server/proxy.js:    var dfd = this.async(1000);
./tests/server/proxy.js:    var dfd = this.async(1000);
./tests/server/proxy.js:    var dfd = this.async(1000);
./tests/server/proxy.js:    var dfd = this.async(1000);
./tests/server/templates.js:      var dfd = this.async(1000);
```

Plus the few places where we did `* 1000` instead of `1000 *`:
```sh
$ grep -Ir "* 1000" . --exclude-dir="node_modules" --exclude-dir=".tmp" --exclude-dir="app/bower_components" --exclude-dir="dist"  | grep -v ".svg:" | grep -v "bower_components" | grep -v app/scripts/vendor/

./app/scripts/lib/metrics.js:  var TEN_MINS_MS = 10 * 60 * 1000;
./app/scripts/views/mixins/resend-mixin.js:  var SHOW_RESEND_IN_MS = 5 * 60 * 1000; // 5 minutes.
./server/lib/configuration.js:    default: 10 * 60 * 1000 // 10 minutes
```

### PR DESCRIPTION
**Pull request by larissagaulia**
fixes #1657: add duration-js to convert durations into a more readable format

### CODE REVIEW

File: bower.json [hunk 1]
```diff
@@ -37,7 +37,8 @@
     "tos-pp": "https://github.com/mozilla/legal-docs.git",
     "underscore": "1.8.3",
     "webrtc-adapter": "0.2.5",
-    "vat": "0.0.5"
+    "vat": "0.0.5",
+    "Duration.js": "duration.js#~3.6.0"
```
Comment (vladikoff) [inline bower.json]: we need to make sure to lock this version of the dependency to avoid issues in the future, so just remove the `~`. (Just like other ones in this file)

File: app/scripts/require_config.js [hunk 2]
```diff
@@ -27,6 +27,7 @@ require.config({
     chai: '../bower_components/chai/chai',
     cocktail: '../bower_components/cocktail/Cocktail',
     crosstab: 'vendor/crosstab',
+    durationJs: '../bower_components/Duration.js/duration',
```
Comment (vladikoff) [inline app/scripts/require_config.js]: we can probably just call it `duration` here and just `require('duration')`, `durationJs` looks a bit ugly.

File: tests/functional/reset_password.js [hunk 3]
```diff
@@ -108,8 +110,7 @@ define([
     name: 'reset password',
 
     beforeEach: function () {
-      // timeout after 90 seconds
-      this.timeout = 90000;
+      this.timeout = new Duration('90s').milliseconds();
```
Comment (vladikoff) [inline tests/functional/reset_password.js]: All these cases bellow `this.timeout = new Duration('90s').milliseconds();` can just use the `TIMEOUT`  value from above.

File: tests/functional/reset_password.js [hunk 4]
```diff
@@ -439,8 +439,7 @@ define([
     },
 
     'complete reset, then re-open verification link, click resend': function () {
-      // timeout after 90 seconds
-      this.timeout = 90000;
+      this.timeout = new Duration('90s').milliseconds();
```
Comment (vladikoff) [inline tests/functional/reset_password.js]: Ref: https://github.com/mozilla/fxa-content-server/pull/3621/files#r57535457

File: tests/functional/reset_password.js [hunk 5]
```diff
@@ -88,7 +90,7 @@ define([
   function testAtSettingsWithVerifiedMessage(context) {
     return context.remote
       .setFindTimeout(intern.config.pageLoadTimeout)
-      .sleep(1000)
+      .sleep(new Duration('1s').milliseconds())
```
Comment (vladikoff) [inline tests/functional/reset_password.js]: You can create a `SLEEP_TIMEOUT` above right after `var TIMEOUT` above.

File: app/tests/spec/views/confirm.js [hunk 6]
```diff
@@ -164,7 +165,7 @@ define(function (require, exports, module) {
           callback();
         });
 
-        view.VERIFICATION_POLL_IN_MS = 100;
+        view.VERIFICATION_POLL_IN_MS = Duration('100ms');
```
Comment (vladikoff) [inline app/tests/spec/views/confirm.js]: missing `new` ?

**Comment by vladikoff** [comment]
@larissagaulia Hey, so after looking through this pull request: 

- Let's remove changes to files in `tests/functional/`, it's a bit tricky and not that useful there.
- Changes in `app/scripts/lib/assertion.js` are good
- ` durationJs: '../bower_components/Duration.js/duration',` let's try to rename that to just `duration`.
- You can still remove repetition in https://github.com/mozilla/fxa-content-server/pull/3621/files#r57535457 if you want. Just make `self.timeout = TIMEOUT` where `var TIMEOUT = 90 * 1000;`

@pdehaan let me know if you agree / disagree.

**Comment by vladikoff** [comment]
@larissagaulia one more tip: in your commit messages instead of 

```
refactor(client): Convert durations into duration-js issue1657
```

You can do 

```
refactor(client): Convert durations into duration-js

Fixes #1657 
```

This will make it more consistent with our commit message style and also close issues you work on automatically. See this page: https://help.github.com/articles/closing-issues-via-commit-messages/

File: app/scripts/lib/assertion.js [hunk 7]
```diff
@@ -5,11 +5,12 @@
...
   'use strict';
 
+  var Duration = require('durationJs');
   var jwcrypto = require('vendor/jwcrypto');
   var P = require('lib/promise');
 
-  var CERT_DURATION_MS = 1000 * 60 * 60 * 6; // 6hrs
-  var ASSERTION_DURATION_MS = 1000 * 3600 * 24 * 365 * 25; // 25 years
+  var CERT_DURATION_MS =  new Duration('6h').milliseconds();
+  var ASSERTION_DURATION_MS = new Duration('52w').milliseconds() * 25; //25 years
```
Comment (vladikoff) [inline app/scripts/lib/assertion.js]: @larissagaulia see if you can find a way to validate that `Duration` for `ASSERTION_DURATION_MS` works correctly. We want to make sure assertion is correctly set to  expire in 25 years.
You can add an assert to existing tests in `app/tests/spec/lib/assertion.js`. 

also see below how it is used (` expiresAt: Date.now() + ASSERTION_DURATION_MS`)

File: tests/intern.js [hunk 8]
```diff
@@ -10,7 +10,9 @@ define([
   'intern/node_modules/dojo/has!host-node?intern/node_modules/dojo/topic',
   './tools/firefox_profile'
 ],
-function (intern, topic, firefoxProfile) {
+function (require, intern, topic, firefoxProfile) {
+  var Duration = require('durationJs');
```
Comment (vladikoff) [inline tests/intern.js]: @larissagaulia undo the changes in `tests/intern.js` and `tests/server/*` just like `tests/functional/*`

File: app/scripts/lib/constants.js [hunk 9]
```diff
@@ -44,7 +45,7 @@ define(function (require, exports, module) {
     SYNC11_MIGRATION: 'sync11',
     AMO_MIGRATION: 'amo',
 
-    VERIFICATION_POLL_IN_MS: 4000,
+    VERIFICATION_POLL_IN_MS: new Duration('400ms').milliseconds(),
```
Comment (vladikoff) [inline app/scripts/lib/constants.js]: @larissagaulia this should be 4 seconds not 400 ms

Comment (larissagaulia) [inline app/scripts/lib/constants.js]: thanks. I'll go over them all again to see if I missed any others.

**Comment by vladikoff** [comment]
Ok that's all the comments from me on this pull request for now. Great work @larissagaulia! Readable durations are going to be awesome!

File: app/scripts/lib/assertion.js (hunk 7 above)
Comment (larissagaulia) [inline app/scripts/lib/assertion.js]: Doesn't 
```
if (assertionComponents.payload.exp < (Date.now() + LONG_LIVED_ASSERTION_DURATION - 5000)) {
      throw new Error('assertion should be long lived');
}
```

cover it? I mean, LONG_LIVED_ASSERTION_DURATION is set to the same amount as ASSERTION_DURATION_MS

Comment (pdehaan) [inline app/scripts/lib/assertion.js]: @vladikoff:

```js
var Duration = require("duration-js")

var ASSERTION_DURATION_MS = 1000 * 3600 * 24 * 365 * 25; // 25 years
var ASSERTION_DURATION_MS2 = new Duration('52w').milliseconds() * 25; //25 years

console.log(ASSERTION_DURATION_MS); // 788400000000
console.log(ASSERTION_DURATION_MS2); // 786240000000
```

788400000000 - 786240000000 === 2160000000 ms or 600 hours or 25 days difference.

**Comment by larissagaulia** [comment]
Hmm, can't figure out why 
`Error: Failed to load module duration from /home/ubuntu/fxa-content-server/duration.js (parent: app/scripts/lib/constants)` on CircleCI

**Comment by pdehaan** [comment]
Travis-CI logs are saying:

```
$ travis_retry npm run test-travis

> fxa-content-server@0.59.0 test-travis /home/travis/build/mozilla/fxa-content-server
> intern-runner config=tests/intern_travis
Creating Firefox profile...
Listening on 0.0.0.0:9090

Error: Failed to load module duration from /home/travis/build/mozilla/fxa-content-server/duration.js (parent: app/scripts/lib/constants)
  at <node_modules/intern/node_modules/dojo/loader.ts:831:119>
  at <fs.js:208:20>
  at Object.oncomplete  <fs.js:108:15>
```

**Comment by jrgm** [comment]
I don't want to muddle the conversation, but I have to note that we use `node-convict` in our repositories, and `node-convict` uses `moment` to express time strings. Given that, wouldn't it be reasonable to use `moment`, to avoid any dissonance?

**Comment by vladikoff** [comment]
@jrgm these are on the front-end. moment is a huge dependency to include on the front-end. duration is like 100 lines unminified.

File: tests/lib/restmail.js [hunk 10]
```diff
@@ -5,7 +5,9 @@
 define([
   'tests/lib/request',
   'intern/node_modules/dojo/Promise'
-], function (request, Promise) {
+], function (require, request, Promise) {
+
+  var Duration = require('duration');
```
Comment (vladikoff) [inline tests/lib/restmail.js]: undo changes in this file and that should unblock circle ci

**Comment by jrgm** [comment]
@vladikoff - thanks for clarifying (I had to ask).

File: app/scripts/lib/constants.js [hunk 11]
```diff
@@ -5,6 +5,7 @@
 define(function (require, exports, module) {
   'use strict';
 
+  var Duration = require('duration');
```
Comment (vladikoff) [inline app/scripts/lib/constants.js]: @larissagaulia looking at the tests failures I see that `app/scripts/lib/constants.js` is also used by server code, which means we won't be able to use Duration here. So you need to undo this change.

File: app/scripts/lib/assertion.js (hunk 7 above)
Comment (vladikoff) [inline app/scripts/lib/assertion.js]: @pdehaan How are you feeling about that 25 day difference? Should we keep the numbers as they are or should we make a lib that calculates the years the same way as moment?

Comment (pdehaan) [inline app/scripts/lib/assertion.js]: I think if we're talking 25 years (2041), I'm personally not too concerned about it being +/-1 month. That said, possibly not a horrible idea to create a simple date lib which just does the 11 lines of code we need from moment-js. Or just merge as-is, and file a new bug to consider in the future.

Comment (jrgm) [inline app/scripts/lib/assertion.js]: As we all know, 32-bit epoch time math is invalid past 2038-01-19. Punt! ;-)

File: app/tests/spec/views/confirm.js [hunk 12]
```diff
@@ -164,7 +165,7 @@ define(function (require, exports, module) {
           callback();
         });
 
-        view.VERIFICATION_POLL_IN_MS = 100;
+        view.VERIFICATION_POLL_IN_MS = new Duration('100ms');
```
Comment (vladikoff) [inline app/tests/spec/views/confirm.js]: @larissagaulia this needs to a `.milliseconds()`

Ref: https://tonicdev.com/vladikoff/56f8a8646f72571100d56ab7

**Comment by vladikoff** [comment]
@larissagaulia good stuff! Just one issue left: https://github.com/mozilla/fxa-content-server/pull/3621#discussion_r57542850

**Comment by vladikoff** [comment]
@larissagaulia awesome work, Thank you!

**Comment by pdehaan** [comment]
:+1: A big thank you to @larissagaulia for getting this finished!

**Comment by larissagaulia** [comment]
My pleasure!
